### DISCUSSION TITLE
Implemented Viem in our codebase but struggling a lot with typescript

### DISCUSSION THREAD

**Discussion by NBMSacha** [wevm/viem#1609] [Question]
GM!

We're been implementing viem to [Mobula's front-end](https://github.com/MobulaFi/mobula-ui) & we add many hacky stuff to do to get it compile with typescript, here's an example: 

```
 const publicClient = createPublicClient({
            chain: idToWagmiChain[blockchain.chainId],
            transport: http(blockchain.rpcs[0]),
          });

 const contract = getContract({
            address: search as `0x${string}`,
            abi: erc20ABI,
            publicClient: publicClient as never,
}) as any;

const symbol = await contract.read.symbol();
```

If we don't add "as never"  and "as any", we get: 

```
Type '{ account: undefined; batch?: { multicall?: boolean | { batchSize?: number; wait?: number; }; }; cacheTime: number; chain: Chain; key: string; name: string; pollingInterval: number; request: EIP1193RequestFn<PublicRpcSchema>; ... 49 more ...; extend: <const client extends { ...; }>(fn: (client: Client<...>) => clien...' is not assignable to type 'never'.
  The intersection 'Client<Transport, Chain>' was reduced to 'never' because property 'cacheTime' has conflicting types in some constituents.t
```

And 

`Property 'read' does not exist on type 'never'.t` 

Which might be the same issue.

Any idea what we did wrong?

**Comment by tmm** [comment]
Do you have TypeScript strict mode on? Can you share a complete example for what you are running into?

**Reply by metabacalhau** [reply]
Hi @tmm, here is the code that gives me the error:

```
import { Chain, createPublicClient, getContract, http } from 'viem'

export async function getWeb3ViemClients() {
  const publicClient = createPublicClient({
    chain: {} as Chain,
    transport: http(),
  })

  getContract({
    address: '0x0000000000000000000000000000000000000000',
    client: {
      public: publicClient,
    },
  })
}
```

The error happens at line `public: publicClient,`:
`Type '{ account: undefined; batch?: { multicall?: boolean | { batchSize?: number; wait?: number; }; }; cacheTime: number; chain: Chain; key: string; name: string; pollingInterval: number; request: EIP1193RequestFn<PublicRpcSchema>; ... 50 more ...; extend: <const client extends { ...; } & ExactPartial<...>>(fn: (client: C...' is not assignable to type 'never'.
  The intersection 'Client<Transport, Chain>' was reduced to 'never' because property 'cacheTime' has conflicting types in some constituents.`

TypeScript: 5.4.2
Viem: 2.8.13

Could you please tell us how to fix this issue?

**Reply by tmm** [reply]
Works fine for me: [TypeScript Playground](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAbzgYQBYENgDsA0cDGUApujEQAoCuARgDbD7L1FYx4DmRMyErU6+NnFQwYYPESj4ATAAYAgtWBwAvnABmUCCDgByAG7AiIXQChTRAB6RYcdAGcAnlnwbKLmMF5xOMAOpE1ADMAGpGIExGrPYAFACUiKZwBLz28GA09IzMrHAAvATEpBSZDJEsMDEISckEGNgAXIhqDij1uDXJMPxY9jYwTSJi8Tg1KnHmyb48fAKV1bV2ACZLxPb2TbqylrK7e-sHh0fH+7qji+hKTZIyCkrntfg5A4mLyRl0DE0fWeWsD8kVA9xqYVKYgA)

**Comment by tungkiet7** [comment]
@tmm Could you please tell us how to fix this issue?

**Reply by DLHTX** [reply]
I have the same problem. You can check whether the tsconfig file has StructNullChecks, and if not,just add `"StructNullChecks": true.`

**Reply by tinwritescode** [reply]
@DLHTX Thanks. I fixed it but with `"strictNullChecks": true` instead of StructNullChecks.

**Reply by junhoyeo** [reply]
-> `strictNullChecks`

**Comment by SanyaSemenov** [comment]
@NBMSacha please tell that you managed to handle it 😅 please provide your solution if you have one

**Reply by NBMSacha** [reply]
Yes, I was a typescript newbie at the time, that's why. The way to fix this is so use "as const" on your ABIs & most stuff.

**Reply by SanyaSemenov** [reply]
Nah, I wish the issue was about it, but I tend to think that it's packages versions mismatch because even if I copypaste pancakeswap's [getContract helper](https://github.com/pancakeswap/pancake-frontend/blob/develop/apps/web/src/utils/contractHelpers.ts) (I'm making smth similar), the issue is the same (about 'cacheTime')
